tools/testbench/file.c

```diff
@@ -683,16 +683,86 @@ static int file_verify_params(struct comp_dev *dev,
 static int file_params(struct comp_dev *dev,
 		       struct sof_ipc_stream_params *params)
 {
-	int err;
+	struct comp_buffer *buffer = NULL;
+	struct dai_data *dd = comp_get_drvdata(dev);
+	struct file_comp_data *cd = comp_get_drvdata(dd->dai);
+	struct audio_stream *stream;
+	int periods;
+	int samples;
+	int ret;
 
 	comp_info(dev, "file_params()");
 
-	err = file_verify_params(dev, params);
-	if (err < 0) {
+	ret = file_verify_params(dev, params);
+	if (ret < 0) {
 		comp_err(dev, "file_params(): pcm params verification failed.");
+		return ret;
+	}
+
+	/* file component sink/source buffer period count */
+	switch (cd->fs.mode) {
+	case FILE_READ:
+		buffer = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+		periods = dev->ipc_config.periods_sink;
+		break;
+	case FILE_WRITE:
+		buffer = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+		periods = dev->ipc_config.periods_source;
+		break;
+	default:
+		/* TODO: duplex mode */
+		fprintf(stderr, "Error: Unknown file mode %d\n", cd->fs.mode);
+		break;
+	}
+
+	if (!buffer) {
+		fprintf(stderr, "error: no sink/source buffer\n");
+		return -EINVAL;
+	}
```

Comment:
you could just put this `fprintf(); return -EINVAL;` under `default:` above - that's the only case when this is entered. Then you also wouldn't need to initialise `buffer`

Reply:
What if list_first_item() would really return NULL buffer with some broken topology? Do you think such is not possible.

Reply:
in our list implementation `list_first_item()` (almost) never returns NULL - as long as the list is valid. For an empty list it will return the address of the list header possibly minus offset of that list header in listed objects

Reply:
This is now fixed.